kmath-core/src/commonMain/kotlin/space/kscience/kmath/misc/sorting.kt

```diff
@@ -0,0 +1,99 @@
+ * permutations (shuffle, sort), etc.
+ *
+ * Defining an API for this could help factorizing code, and more importantly, define a unified set of properties/operators over 
+ * a wide range of data structures. 
+ */
+private interface Registry<K, V> {
+    fun keys() : Sequence<K>
+    operator fun get(key: K) : V? 
+}
+
+private interface MonotonicIntRegistry<V> : Registry<Int, V> {
+    val indices : IntRange
+    override fun keys() = indices.asSequence()
+}
+
+private fun <V> Buffer<V>.subset(interval: IntRange) : MonotonicIntRegistry<V> {
+    require(0 <= interval.start && interval.start < size) { "Interval is out of bounds: invalid start" }
+    require(0 <= interval.last && interval.last < size) { "Interval is out of bounds: invalid end" }
+    val buf = this
+    return object : MonotonicIntRegistry<V> {
+        override operator fun get(key: Int) = buf.get(key)
+        override val indices = interval
+    }
+}
+
+private fun <V: Comparable<V>> MonotonicIntRegistry<V>.permSort(descending : Boolean = false) : IntArray {
+    val offset = indices.start
+    val stride = indices.step
+    val size = (indices.last - offset + stride) / stride
+    if (size == 0) return IntArray(0)
+    else if (size == 1) return intArrayOf(offset)
+
+    val comparator = if (descending) compareByDescending<Int> { get(it) } else compareBy<Int> { get(it) }
+
+    /* TODO: optimisation when stride is 1 and offset is 0: keep a constant big array of indices (Ex, from 0 to 4096),
+     * then create indice arrays more efficiently by copying subpart of cached one. For bigger needs, we could copy 
+     * entire cached array, then fill remaining indices manually.
+     * Not done for now, because:
+     *  1. doing it right would require some statistics about common used buffer sizes.
+     *  2. Some benchmark would be needed to ensure it would really provide better performance
+     */
+    val packedIndices = IntArray(size) { idx -> offset + idx * stride }
+    
+    /* TODO: find an efficient way to sort in-place instead, and return directly the IntArray. 
+     * Not done for now, because no standard utility is provided yet. An open issue exists for this.
+     * See: https://youtrack.jetbrains.com/issue/KT-37860 
+     */
+    return packedIndices.sortedWith(comparator).toIntArray()
+}
+
+/**
+ * Return a new list filled with buffer indices. Indice order is defined by sorting associated buffer value.
+ * This feature allows to sort buffer values without reordering its content.
+ * 
+ * @param descending True to revert sort order from highest to lowest values. Default to ascending order. 
+ * @param offset Index to start sorting from input buffer. Default to the start of the buffer.
+ * @param limit Number of elements to sort. Default to remaining elements after input offset.
+ * @return List of buffer indices, sorted by associated value.
+ */
+@UnstableKMathAPI
+public fun <V : Comparable<V>> Buffer<V>.permSort(
```

Comment:
I do not think we need offsets here. We have (or will have) buffer view functionality for that.